### PR TITLE
[4.0] Fix Issue  Session timer not working.

### PR DESCRIPTION
**Pull request by anuragteapot**
Pull Request for Issue #18921  

### Summary of Changes
Changes in :
<code> \joomla-cms\libraries\vendor\joomla\session\src\Session.php</code><br>
 In Start a session and  Restart an expired or locked session.
 
Added : Validate session if the session isn't valid and if it expired then try to restart  else destroy it.

### Testing Instructions
Apply patch  and do nothing for period of time  you set in session timer (lifetime).

### Expected result
auto logout.

### Actual result
still active.

### CODE REVIEW

**Comment by mbabker** [comment]
This PR needs to be submitted against https://github.com/joomla-framework/session/tree/2.0-dev

The `libraries/vendor` directory is all files from external repositories as compared to the CMS repo.

**Comment by anuragteapot** [comment]
@mbabker But , this file is related to Dev 4.0 branch.

**Comment by mbabker** [comment]
Understand completely.  But, the file you've changed is technically a third party dependency to this repository (as all files in `libraries/vendor` are).  So the change needs to be applied to the source repository otherwise the next time a `composer update` is run on this repository the change will be overwritten.

File: libraries/vendor/joomla/session/src/Session.php [hunk 1]
```diff
@@ -464,7 +464,20 @@ public function start()
...
+			// else destroy it.
+			if ($this->state === 'expired')
+			{
+				$this->restart();
+			}
+			else
+			{
+				$this->destroy();
+			}
+
```
Comment (Quy) [inline libraries/vendor/joomla/session/src/Session.php:479]: Remove blank line.

File: libraries/vendor/joomla/session/src/Session.php [hunk 2]
```diff
@@ -523,17 +536,26 @@ public function restart()
...
 		$this->store->start();
 
 		$this->setCounter();
 		$this->setTimers();
-		$this->validate(true);
-
-		// Restore the data
-		foreach ($data as $key => $value)
+		
+
```
Comment (Quy) [inline libraries/vendor/joomla/session/src/Session.php:551]: Remove line and tabs.

File: libraries/vendor/joomla/session/src/Session.php [hunk 3]
```diff
@@ -523,17 +536,26 @@ public function restart()
...
-		$this->validate(true);
-
-		// Restore the data
-		foreach ($data as $key => $value)
+		
+		
+		if(!$this->validate(true))
 		{
-			$this->set($key, $value);
+			/**
```
Comment (Quy) [inline libraries/vendor/joomla/session/src/Session.php:554]: Replace with single inline comment.
`// Destroy the session if it's not valid.`

**Comment by mbabker** [comment]
Superseded by https://github.com/joomla-framework/session/pull/34

**Comment by ghost** [comment]
@mbabker close #18921?

**Comment by mbabker** [comment]
No, issue should stay open here until the PR is merged upstream and then the fix applied to this repo as well.

**Pull request closed without merging**